### PR TITLE
Fix #9831 - Issue running a model with PIU + ReturnPlenu

### LINKED ISSUE #9831: Issue running a model with PIU + ReturnPlenum
Issue overview
--------------

When a return plenum is used in conjunction with a PIU, the `PoweredInductionUnits::PIUInducesPlenumAir` is called here in `ZonePlenum::GetZonePlenumInput`: https://github.com/NREL/EnergyPlus/blob/4a7322f8f2005215913c795504b4da46b5937086/src/EnergyPlus/ZonePlenum.cc#L423

That in turns calls `PoweredInductionUnits::GetPIUs`, but at that point in time, `ZoneAirLoopEquipmentManager::GetZoneAirLoopEquipment` has not been called yet and therefore it ends up issuing a fatal that it can't find the ADU:

https://github.com/NREL/EnergyPlus/blob/4a7322f8f2005215913c795504b4da46b5937086/src/EnergyPlus/PoweredInductionUnits.cc#L479-L495

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): all
 - Version of EnergyPlus (if using an intermediate build, include SHA): 9.6, 22.2.0, develop
 - reported by @eringold

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added: https://github.com/NREL/EnergyPlusDevSupport/blob/master/DefectFiles/9000s/9831/in.idf
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #9831
- Refactor PoweredInductionUnits.cc and partially ZonePlenum.cc

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/PoweredInductionUnits.unit.cc [hunk 1]
```diff
@@ -792,3 +809,1170 @@ TEST_F(EnergyPlusFixture, SeriesPIUZoneOAVolumeFlowRateTest)
...
+        "  SeriesPIU Secondary Air Inlet Node,     !- Secondary Air Inlet Node Name",
+        "  SeriesPIU Outlet Node,                  !- Outlet Node Name",
+        "  Air Terminal Single Duct Series PIU Reheat 1 Fan Outlet, !- Reheat Coil Air Inlet Node Name",
+        "  Air Terminal Single Duct Series PIU Reheat 1 Mixer, !- Zone Mixer Name",
+        "  Fan System Model 1,                     !- Fan Name",
+        "  Coil:Heating:Electric,                  !- Reheat Coil Object Type",
+        "  Coil Heating Electric 2,                !- Reheat Coil Name",
+        "  Autosize,                               !- Maximum Hot Water or Steam Flow Rate {m3/s}",
+        "  0,                                      !- Minimum Hot Water or Steam Flow Rate {m3/s}",
+        "  0.001;                                  !- Convergence Tolerance",
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/PoweredInductionUnits.unit.cc:1270]: Unit test: The PIU does have a correctly defined ADU.

File: tst/EnergyPlus/unit/PoweredInductionUnits.unit.cc [hunk 2]
```diff
@@ -792,3 +809,1170 @@ TEST_F(EnergyPlusFixture, SeriesPIUZoneOAVolumeFlowRateTest)
...
+        "  AirLoopHVAC:ZoneMixer,                  !- Component Object Type 2",
+        "  Air Loop HVAC Zone Mixer 1;             !- Component Name 2",
+
+        "AirLoopHVAC:ReturnPlenum,",
+        "  Air Loop HVAC Return Plenum 1,          !- Name",
+        "  ReturnPlenum,                           !- Zone Name",
+        "  ReturnPlenum Zone Air Node,             !- Zone Node Name",
+        "  Plenum Outlet Node,                     !- Outlet Node Name",
+        "  SeriesPIU Secondary Air Inlet Node,     !- Induced Air Outlet Node or NodeList Name",
+        "  Zone1 Return Air Node;                  !- Inlet Node Name 1",
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/PoweredInductionUnits.unit.cc:1874]: You need a return plenum, with an Induced Air Outlet Node defined.

File: tst/EnergyPlus/unit/PoweredInductionUnits.unit.cc [hunk 3]
```diff
@@ -792,3 +809,1170 @@ TEST_F(EnergyPlusFixture, SeriesPIUZoneOAVolumeFlowRateTest)
...
+        "  Air Loop HVAC Zone Splitter 1,          !- Name",
+        "  Node 3,                                 !- Inlet Node Name",
+        "  Node 12;                                !- Outlet Node Name 1",
+
+        "AirLoopHVAC:SupplyPlenum,",
+        "  Air Loop HVAC Supply Plenum 1,          !- Name",
+        "  SupplyPlenum,                           !- Zone Name",
+        "  SupplyPlenum Zone Air Node,             !- Zone Node Name",
+        "  Node 12,                                !- Inlet Node Name",
+        "  SeriesPIU Supply Air Inlet Node;        !- Outlet Node Name 1",
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/PoweredInductionUnits.unit.cc:1858]: You also need a Supply Plenum to trigger the issue.

File: tst/EnergyPlus/unit/PoweredInductionUnits.unit.cc [hunk 4]
```diff
@@ -792,3 +809,1170 @@ TEST_F(EnergyPlusFixture, SeriesPIUZoneOAVolumeFlowRateTest)
...
+    state->dataGlobal->BeginEnvrnFlag = true;
+    state->dataGlobal->ZoneSizingCalc = true;
+    EXPECT_FALSE(has_err_output(true));
+    EXPECT_NO_THROW(SizingManager::ManageSizing(*state));
+
+    std::string const expectedError = delimited_string({
+        "   ************* Beginning Zone Sizing Calculations",
+        "   ************* Beginning System Sizing Calculations",
+    });
+    EXPECT_TRUE(compare_err_stream(expectedError, true));
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/PoweredInductionUnits.unit.cc:1977]: Before fix:

```diff
+   ** Severe  ** GetPIUs: No matching Air Distribution Unit, for PIU = [AirTerminal:SingleDuct:SeriesPIU:Reheat,AIR TERMINAL SINGLE DUCT SERIES PIU REHEAT 1].
+   **   ~~~   ** ...should have outlet node = SERIESPIU OUTLET NODE
+   **  Fatal  ** GetPIUs: Errors found in getting input.  Preceding conditions cause termination.
+   ...Summary of Errors that led to program termination:
+   ..... Reference severe error count=1
+   ..... Last severe error=GetPIUs: No matching Air Distribution Unit, for PIU = [AirTerminal:SingleDuct:SeriesPIU:Reheat,AIR TERMINAL SINGLE DUCT SERIES PIU REHEAT 1].\n
```

File: src/EnergyPlus/PoweredInductionUnits.cc [hunk 5]
```diff
@@ -262,6 +263,15 @@ void GetPIUs(EnergyPlusData &state)
...
         state.dataInputProcessing->inputProcessor->getNumObjectsFound(state, "AirTerminal:SingleDuct:ParallelPIU:Reheat");
     state.dataPowerInductionUnits->NumPIUs = state.dataPowerInductionUnits->NumSeriesPIUs + state.dataPowerInductionUnits->NumParallelPIUs;
+
+    if (state.dataPowerInductionUnits->NumPIUs > 0) {
+        // GetZonePlenumInput might call this routine before the AirDistUnit has been populated
+        if (state.dataZoneAirLoopEquipmentManager->GetAirDistUnitsFlag) {
+            ZoneAirLoopEquipmentManager::GetZoneAirLoopEquipment(state);
+            state.dataZoneAirLoopEquipmentManager->GetAirDistUnitsFlag = false;
+        }
+    }
```
Comment (jmarrec) [inline src/EnergyPlus/PoweredInductionUnits.cc:263]: Atop of GetPIUs, cann ZoneAirLoopEquipmentManager::GetZoneAirLoopEquipment so we can check the ADU.

File: src/EnergyPlus/PoweredInductionUnits.cc [hunk 6]
```diff
@@ -285,23 +285,22 @@ void GetPIUs(EnergyPlusData &state)
...
                                                  ErrorsFound);
-        state.dataPowerInductionUnits->PIU(PIUNum).Name = state.dataIPShortCut->cAlphaArgs(1);
-        state.dataPowerInductionUnits->PIU(PIUNum).UnitType = cCurrentModuleObject;
-        state.dataPowerInductionUnits->PIU(PIUNum).UnitType_Num = DataDefineEquip::ZnAirLoopEquipType::SingleDuct_SeriesPIU_Reheat;
-        state.dataPowerInductionUnits->PIU(PIUNum).Sched = state.dataIPShortCut->cAlphaArgs(2);
+        auto &thisPIU = state.dataPowerInductionUnits->PIU(PIUIndex);
+        thisPIU.Name = state.dataIPShortCut->cAlphaArgs(1);
+        thisPIU.UnitType = cCurrentModuleObject;
+        thisPIU.UnitType_Num = DataDefineEquip::ZnAirLoopEquipType::SingleDuct_SeriesPIU_Reheat;
+        thisPIU.Sched = state.dataIPShortCut->cAlphaArgs(2);
```
Comment (jmarrec) [inline src/EnergyPlus/PoweredInductionUnits.cc:298]: The rest of changes are about taking references and limiting scope of variables (and initializing them)

**Comment by Myoldmopar** [comment]
Small conflict here, could resolve on GitHub but I'd still need to test and such so I'll just do that locally now.

**Comment by Myoldmopar** [comment]
Pulled develop in here and built/tested.  Simple merge conflict just due to the TimeStepSys cleanup.  I don't think this needs to wait until CI is done, it was 100% all super green before.  I'm just getting this one merged in.  Thanks @jmarrec

**Pull request merged**